### PR TITLE
Support getOccurrences on the TS Server

### PR DESCRIPTION
**Pull request by DanielRosenwasser**

### CODE REVIEW

File: src/server/session.ts [hunk 1]
```diff
@@ -89,6 +89,7 @@ module ts.server {
         export var Geterr = "geterr";
         export var NavBar = "navbar";
         export var Navto = "navto";
+        export var Occurrences = "occurrences";
```
Comment (mhegazy) [inline src/server/session.ts]: looks like you are missing the actual handling of the command in onMessage

File: tests/cases/fourslash/server/completions02.ts [hunk 2]
```diff
@@ -1,4 +1,4 @@
-/// <reference path="fourslash.ts" />
+/// <reference path="../fourslash.ts"/>
```
Comment (mhegazy) [inline tests/cases/fourslash/server/completions02.ts]: thanks :)

**Comment by steveluc** [comment]
Get occurences and get references share the same return type.  It would be good if they shared a return type in protocol.d.ts also.

**Comment by steveluc** [comment]
You could make optional the lineText field in the ReferencesResponseItem and then use it for occurrences.

File: src/server/session.ts (hunk 1 above)
Comment (DanielRosenwasser) [inline src/server/session.ts]: Yup, just sent that part out. I'm not that familiar with the server so I'm still getting a feel for things.

**Comment by DanielRosenwasser** [comment]
On one hand, I agree that they are frankly similar and we'd be reusing code; on the other, I somewhat feel like we'd be conflating two concepts. Is `lineText` *always* included for a `ReferencesResponseItem`?

File: tests/cases/fourslash/server/occurrences01.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+////}
+
+let ranges = test.ranges();
+
+for (let r of ranges) {
+    goTo.position(r.start);
+    verify.occurrencesAtPositionCount(ranges.length);
+
+    for (let range of ranges) {
+        verify.occurrencesAtPositionContains(range, false);
```
Comment (CyrusNajmabadi) [inline tests/cases/fourslash/server/occurrences01.ts]: what does 'false' mean?

File: src/server/session.ts [hunk 4]
```diff
@@ -89,6 +89,7 @@ module ts.server {
         export var Geterr = "geterr";
         export var NavBar = "navbar";
         export var Navto = "navto";
+        export var Occurrences = "occurrences";
         export var Open = "open";
         export var Quickinfo = "quickinfo";
```
Comment (CyrusNajmabadi) [inline src/server/session.ts]: I feel like a lot of these names need cleaning up.  Certain names match VS, but do not match the casing there (like Quickinfo vs. QuickInfo).  Other names are inconsistently capitalized ("NavBar" vs "Navto").  Some names are abbreviated, some are not.   Some names match Roslyn, some do not ("GetOccurrences" vs "GetHighlights").

Comment (DanielRosenwasser) [inline src/server/session.ts]: :+1: I'd support fixing this up, but in a different PR.

File: src/server/session.ts [hunk 5]
```diff
@@ -884,6 +915,11 @@ module ts.server {
                         response = this.getNavigationBarItems(navBarArgs.file);
                         break;
                     }
+                    case CommandNames.Occurrences: {
+                        var { line, offset, file: fileName } = <protocol.FileLocationRequestArgs>request.arguments;
```
Comment (mhegazy) [inline src/server/session.ts]: if you are using destructuring i would put it everywhere in this method for consistency.

**Comment by mhegazy** [comment]
:+1:

**Pull request merged**